File: src/cmds/git.ts
import { run } from '../core/runner';
import { stripAnsi, truncate } from '../core/utils';

export interface HandlerResult {
  filteredOutput: string;
  exitCode: number;
  rawOutput: string;
  cmdType: string;
  execMs: number;
}

export function handleGit(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || 'status';
  const result = run('git', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  let cmdType = `git ${sub}`;

  try {
    switch (sub) {
      case 'status':
      case 'st':
        filtered = formatStatusOutput(result.stdout, ultra);
        break;
      case 'diff':
        filtered = compactDiff(result.stdout, ultra);
        break;
      case 'log':
        filtered = formatLog(result.stdout, ultra);
        break;
      case 'push':
        filtered = compactPush(raw);
        break;
      case 'pull':
        filtered = compactPull(raw);
        break;
      case 'add':
        filtered = ultra ? '✓' : 'ok';
        if (result.exitCode !== 0) filtered = (result.stderr || raw).trim() || filtered;
        break;
      case 'commit':
        filtered = compactCommit(raw, ultra);
        break;
      case 'branch':
        filtered = compactBranch(result.stdout, ultra);
        break;
      case 'fetch':
        filtered = ultra ? '✓ fetch' : 'ok: fetched';
        break;
      default:
        filtered = truncate(stripAnsi(raw).trim(), 50);
    }
  } catch {
    filtered = raw;
  }

  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : (result.stderr || raw).trim()),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType,
    execMs: result.execMs,
  };
}

function formatStatusOutput(raw: string, ultra = false): string {
  const clean = stripAnsi(raw);
  if (/working tree clean/i.test(clean) || clean.trim() === '') {
    return ultra ? 'clean' : 'nothing to commit, working tree clean';
  }

  const counts: Record<string, number> = {
    modified: 0,
    'new file': 0,
    deleted: 0,
    renamed: 0,
    untracked: 0,
  };

  for (const line of clean.split('\n')) {
    const m = /^\s*(modified|new file|deleted|renamed):\s+(.+)/.exec(line);
    if (m) {
      counts[m[1]]++;
      continue;
    }
    if (/^\s+([^\s].*)$/.test(line) && /Untracked files/i.test(clean)) {
      // skip headers
    }
    if (/^\?\?\s+/.test(line)) {
      counts.untracked++;
    }
  }

  // Also parse porcelain-like lines (M, A, D, R, ??)
  const porcelain = clean.split('\n').filter((l) => /^[ MADRCU?][ MADRCU?]\s/.test(l));
  if (porcelain.length > 0 && counts.modified + counts['new file'] + counts.deleted + counts.untracked === 0) {
    for (const line of porcelain) {
      const code = line.slice(0, 2);
      if (code === '??') counts.untracked++;
      else if (/[M]/.test(code)) counts.modified++;
      else if (/[A]/.test(code)) counts['new file']++;
      else if (/[D]/.test(code)) counts.deleted++;
      else if (/[R]/.test(code)) counts.renamed++;
    }
  }

  // Untracked from "Untracked files:" section
  const untrackedSection = /Untracked files:[\s\S]*?(?=\n\n|\nChanges|$)/.exec(clean);
  if (untrackedSection) {
    const lines = untrackedSection[0]
      .split('\n')
      .filter((l) => /^\t/.test(l) || /^\s{2,}\S/.test(l))
      .filter((l) => !/\(use /.test(l));
    counts.untracked = Math.max(counts.untracked, lines.length);
  }

  if (ultra) {
    const parts: string[] = [];
    if (counts.modified) parts.push(`${counts.modified}M`);
    if (counts['new file']) parts.push(`${counts['new file']}N`);
    if (counts.deleted) parts.push(`${counts.deleted}D`);
    if (counts.renamed) parts.push(`${counts.renamed}R`);
    if (counts.untracked) parts.push(`${counts.untracked}U`);
    return parts.length ? parts.join(' ') : 'clean';
  }

  const parts: string[] = [];
  if (counts.modified) parts.push(`${counts.modified} modified`);
  if (counts['new file']) parts.push(`${counts['new file']} new file`);
  if (counts.deleted) parts.push(`${counts.deleted} deleted`);
  if (counts.renamed) parts.push(`${counts.renamed} renamed`);
  if (counts.untracked) parts.push(`${counts.untracked} untracked`);
  return parts.length ? parts.join(', ') : 'nothing to commit, working tree clean';
}

function compactDiff(raw: string, ultra = false): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n');
  let files = 0;
  let added = 0;
  let removed = 0;
  for (const line of lines) {
    if (line.startsWith('+++ ')) files++;
    else if (line.startsWith('+') && !line.startsWith('+++')) added++;
    else if (line.startsWith('-') && !line.startsWith('---')) removed++;
  }
  if (files === 0 && added === 0 && removed === 0) {
    return ultra ? '0' : 'no changes';
  }
  if (ultra) return `+${added}-${removed}/${files}f`;
  return `${files} file${files === 1 ? '' : 's'}: +${added}/-${removed}`;
}

function formatLog(raw: string, ultra = false): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n');
  const out: string[] = [];
  let current: { hash?: string; subject?: string; date?: string } = {};

  for (const line of lines) {
    const m = /^commit\s+([0-9a-f]+)/.exec(line);
    if (m) {
      if (current.hash) out.push(formatLogEntry(current, ultra));
      current = { hash: m[1].slice(0, 7) };
      continue;
    }
    const dm = /^Date:\s+(.+)/.exec(line);
    if (dm) {
      current.date = dm[1].trim();
      continue;
    }
    if (current.hash && !current.subject) {
      const subject = line.trim();
      if (subject && !/^Author:|^Date:|^Merge:/.test(line)) {
        current.subject = subject;
      }
    }
  }
  if (current.hash) out.push(formatLogEntry(current, ultra));

  if (out.length === 0) {
    // Maybe oneline format
    for (const line of lines) {
      const m = /^([0-9a-f]{7,40})\s+(.+)/.exec(line.trim());
      if (m) {
        out.push(`${m[1].slice(0, 7)} ${m[2]}`);
      }
    }
  }
  return out.join('\n');
}

function formatLogEntry(
  e: { hash?: string; subject?: string; date?: string },
  ultra: boolean,
): string {
  const subject = e.subject || '';
  if (ultra) return `${e.hash} ${subject}`;
  const rel = e.date ? ` (${shortRelative(e.date)})` : '';
  return `${e.hash} ${subject}${rel}`;
}

function shortRelative(iso: string): string {
  const t = new Date(iso).getTime();
  if (isNaN(t)) return iso;
  const ms = Date.now() - t;
  const sec = Math.floor(ms / 1000);
  if (sec < 60) return `${sec}s ago`;
  const min = Math.floor(sec / 60);
  if (min < 60) return `${min}m ago`;
  const hr = Math.floor(min / 60);
  if (hr < 24) return `${hr}h ago`;
  const d = Math.floor(hr / 24);
  if (d < 30) return `${d}d ago`;
  const mo = Math.floor(d / 30);
  if (mo < 12) return `${mo}mo ago`;
  return `${Math.floor(mo / 12)}y ago`;
}

function compactPush(raw: string): string {
  const clean = stripAnsi(raw);
  const m = /\b([\w./-]+)\s*->\s*([\w./-]+)/.exec(clean);
  if (m) return `ok ${m[1]}→${m[2]}`;
  if (/up-to-date|Everything up-to-date/i.test(clean)) return 'ok: up-to-date';
  return 'ok';
}

function compactPull(raw: string): string {
  const clean = stripAnsi(raw);
  if (/Already up.to.date/i.test(clean)) return 'ok: up-to-date';
  const stat = /(\d+)\s+files?\s+changed.*?(\d+)\s+insertions?.*?(\d+)\s+deletions?/.exec(clean);
  const commits = /Fast-forward|Merging|(\d+)\s+commit/.exec(clean);
  const cm = /(\d+)\s+commit/.exec(clean);
  if (stat) {
    const n = cm ? cm[1] : '?';
    return `ok: ${n} commits +${stat[2]}-${stat[3]}`;
  }
  if (commits) return 'ok: pulled';
  return 'ok';
}

function compactCommit(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const m = /\[\S+\s+([0-9a-f]+)\]\s*(.*)/.exec(clean);
  if (m) {
    if (ultra) return `✓ ${m[1].slice(0, 7)}`;
    return `ok ${m[1].slice(0, 7)}: ${m[2]}`;
  }
  if (/nothing to commit/i.test(clean)) return ultra ? 'clean' : 'nothing to commit';
  return ultra ? '✓' : 'ok';
}

function compactBranch(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => l.trim() !== '');
  let current = '';
  const others: string[] = [];
  for (const line of lines) {
    const m = /^\*\s+(.+)/.exec(line);
    if (m) {
      current = m[1].trim();
      continue;
    }
    const t = line.trim();
    if (t) others.push(t);
  }
  if (ultra) return `*${current} (${lines.length}b)`;
  return `current: ${current} | ${lines.length} branches total`;
}

File: internal/handlers/git.go
package handlers

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

func HandleGit(args []string, ultra bool) Result {
	sub := "status"
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("git", args)
	raw := combined(r)
	filtered := ""

	switch sub {
	case "status", "st":
		filtered = formatGitStatus(r.Stdout, ultra)
	case "diff":
		filtered = compactDiff(r.Stdout, ultra)
	case "log":
		filtered = formatGitLog(r.Stdout, ultra)
	case "push":
		filtered = compactPush(raw)
	case "pull":
		filtered = compactPull(raw)
	case "add":
		filtered = "ok"
		if ultra {
			filtered = "✓"
		}
		if r.ExitCode != 0 {
			if s := strings.TrimSpace(firstNonEmpty(r.Stderr, raw)); s != "" {
				filtered = s
			}
		}
	case "commit":
		filtered = compactCommit(raw, ultra)
	case "branch":
		filtered = compactBranch(r.Stdout, ultra)
	case "fetch":
		if ultra {
			filtered = "✓ fetch"
		} else {
			filtered = "ok: fetched"
		}
	default:
		filtered = util.Truncate(strings.TrimSpace(util.StripAnsi(raw)), 50)
	}

	return finalize(filtered, r, raw, "git "+sub)
}

var (
	gitTreeCleanRe    = regexp.MustCompile(`(?i)working tree clean`)
	gitStatusChangeRe = regexp.MustCompile(`^\s*(modified|new file|deleted|renamed):\s+(.+)`)
	gitPorcelainRe    = regexp.MustCompile(`^[ MADRCU?][ MADRCU?]\s`)
	gitUntrackedRe    = regexp.MustCompile(`^\?\?\s+`)
	gitIndentedRe     = regexp.MustCompile(`^\s{2,}\S`)
)

func formatGitStatus(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	if gitTreeCleanRe.MatchString(clean) || strings.TrimSpace(clean) == "" {
		if ultra {
			return "clean"
		}
		return "nothing to commit, working tree clean"
	}

	counts := map[string]int{"modified": 0, "new file": 0, "deleted": 0, "renamed": 0, "untracked": 0}
	for _, line := range strings.Split(clean, "\n") {
		if m := gitStatusChangeRe.FindStringSubmatch(line); m != nil {
			counts[m[1]]++
			continue
		}
		if gitUntrackedRe.MatchString(line) {
			counts["untracked"]++
		}
	}

	// Porcelain fallback (M/A/D/R/?? codes) when the long form matched nothing.
	var porcelain []string
	for _, l := range strings.Split(clean, "\n") {
		if gitPorcelainRe.MatchString(l) {
			porcelain = append(porcelain, l)
		}
	}
	if len(porcelain) > 0 && counts["modified"]+counts["new file"]+counts["deleted"]+counts["untracked"] == 0 {
		for _, line := range porcelain {
			code := line[:2]
			switch {
			case code == "??":
				counts["untracked"]++
			case strings.ContainsRune(code, 'M'):
				counts["modified"]++
			case strings.ContainsRune(code, 'A'):
				counts["new file"]++
			case strings.ContainsRune(code, 'D'):
				counts["deleted"]++
			case strings.ContainsRune(code, 'R'):
				counts["renamed"]++
			}
		}
	}

	// Count files listed under "Untracked files:". RE2 has no lookahead, so we slice by
	// index up to the next blank line / "Changes" header instead of a lookahead regex.
	if sec := untrackedSection(clean); sec != "" {
		n := 0
		for _, l := range strings.Split(sec, "\n") {
			if (strings.HasPrefix(l, "\t") || gitIndentedRe.MatchString(l)) && !strings.Contains(l, "(use ") {
				n++
			}
		}
		if n > counts["untracked"] {
			counts["untracked"] = n
		}
	}

	type part struct {
		n           int
		short, long string
	}
	order := []part{
		{counts["modified"], "M", "modified"},
		{counts["new file"], "N", "new file"},
		{counts["deleted"], "D", "deleted"},
		{counts["renamed"], "R", "renamed"},
		{counts["untracked"], "U", "untracked"},
	}
	var parts []string
	for _, p := range order {
		if p.n == 0 {
			continue
		}
		if ultra {
			parts = append(parts, fmt.Sprintf("%d%s", p.n, p.short))
		} else {
			parts = append(parts, fmt.Sprintf("%d %s", p.n, p.long))
		}
	}
	if len(parts) == 0 {
		if ultra {
			return "clean"
		}
		return "nothing to commit, working tree clean"
	}
	if ultra {
		return strings.Join(parts, " ")
	}
	return strings.Join(parts, ", ")
}

func untrackedSection(clean string) string {
	idx := strings.Index(clean, "Untracked files:")
	if idx < 0 {
		return ""
	}
	rest := clean[idx:]
	end := len(rest)
	if i := strings.Index(rest, "\n\n"); i >= 0 && i < end {
		end = i
	}
	if i := strings.Index(rest, "\nChanges"); i >= 0 && i < end {
		end = i
	}
	return rest[:end]
}

func compactDiff(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	files, added, removed := 0, 0, 0
	for _, line := range strings.Split(clean, "\n") {
		switch {
		case strings.HasPrefix(line, "+++ "):
			files++
		case strings.HasPrefix(line, "+") && !strings.HasPrefix(line, "+++"):
			added++
		case strings.HasPrefix(line, "-") && !strings.HasPrefix(line, "---"):
			removed++
		}
	}
	if files == 0 && added == 0 && removed == 0 {
		if ultra {
			return "0"
		}
		return "no changes"
	}
	if ultra {
		return fmt.Sprintf("+%d-%d/%df", added, removed, files)
	}
	return fmt.Sprintf("%d %s: +%d/-%d", files, plural(files, "file", "files"), added, removed)
}

var (
	gitCommitRe      = regexp.MustCompile(`^commit\s+([0-9a-f]+)`)
	gitDateRe        = regexp.MustCompile(`^Date:\s+(.+)`)
	gitAuthorMergeRe = regexp.MustCompile(`^Author:|^Date:|^Merge:`)
	gitOnelineRe     = regexp.MustCompile(`^([0-9a-f]{7,40})\s+(.+)`)
)

func formatGitLog(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	lines := strings.Split(clean, "\n")

	type entry struct{ hash, subject, date string }
	var cur entry
	var out []string
	flush := func() {
		if cur.hash != "" {
			out = append(out, formatLogEntry(cur.hash, cur.subject, cur.date, ultra))
		}
	}

	for _, line := range lines {
		if m := gitCommitRe.FindStringSubmatch(line); m != nil {
			flush()
			cur = entry{hash: shortHash7(m[1])}
			continue
		}
		if dm := gitDateRe.FindStringSubmatch(line); dm != nil {
			cur.date = strings.TrimSpace(dm[1])
			continue
		}
		if cur.hash != "" && cur.subject == "" {
			if subject := strings.TrimSpace(line); subject != "" && !gitAuthorMergeRe.MatchString(line) {
				cur.subject = subject
			}
		}
	}
	flush()

	if len(out) == 0 { // maybe --oneline
		for _, line := range lines {
			if m := gitOnelineRe.FindStringSubmatch(strings.TrimSpace(line)); m != nil {
				out = append(out, shortHash7(m[1])+" "+m[2])
			}
		}
	}
	return strings.Join(out, "\n")
}

func formatLogEntry(hash, subject, date string, ultra bool) string {
	if ultra {
		return hash + " " + subject
	}
	if date != "" {
		return hash + " " + subject + " (" + shortRelative(date) + ")"
	}
	return hash + " " + subject
}

func shortHash7(h string) string {
	if len(h) > 7 {
		return h[:7]
	}
	return h
}

// shortRelative renders a git date (or ISO date) as "3d ago". Git's default log date isn't
// RFC3339, so we try its layout first.
func shortRelative(s string) string {
	var t time.Time
	ok := false
	for _, layout := range []string{"Mon Jan _2 15:04:05 2006 -0700", time.RFC3339Nano, time.RFC3339} {
		if parsed, err := time.Parse(layout, s); err == nil {
			t, ok = parsed, true
			break
		}
	}
	if !ok {
		return s
	}
	sec := int(time.Since(t).Seconds())
	switch {
	case sec < 60:
		return fmt.Sprintf("%ds ago", sec)
	case sec < 3600:
		return fmt.Sprintf("%dm ago", sec/60)
	case sec < 86400:
		return fmt.Sprintf("%dh ago", sec/3600)
	}
	d := sec / 86400
	if d < 30 {
		return fmt.Sprintf("%dd ago", d)
	}
	mo := d / 30
	if mo < 12 {
		return fmt.Sprintf("%dmo ago", mo)
	}
	return fmt.Sprintf("%dy ago", mo/12)
}

var (
	pushRefRe    = regexp.MustCompile(`\b([\w./-]+)\s*->\s*([\w./-]+)`)
	pushUpToDate = regexp.MustCompile(`(?i)up-to-date|Everything up-to-date`)
)

func compactPush(raw string) string {
	clean := util.StripAnsi(raw)
	if m := pushRefRe.FindStringSubmatch(clean); m != nil {
		return "ok " + m[1] + "→" + m[2]
	}
	if pushUpToDate.MatchString(clean) {
		return "ok: up-to-date"
	}
	return "ok"
}

var (
	pullUpToDateRe = regexp.MustCompile(`(?i)Already up.to.date`)
	pullStatRe     = regexp.MustCompile(`(\d+)\s+files?\s+changed.*?(\d+)\s+insertions?.*?(\d+)\s+deletions?`)
	pullCommitsRe  = regexp.MustCompile(`Fast-forward|Merging|(\d+)\s+commit`)
	pullCommitNRe  = regexp.MustCompile(`(\d+)\s+commit`)
)

func compactPull(raw string) string {
	clean := util.StripAnsi(raw)
	if pullUpToDateRe.MatchString(clean) {
		return "ok: up-to-date"
	}
	if stat := pullStatRe.FindStringSubmatch(clean); stat != nil {
		n := "?"
		if cm := pullCommitNRe.FindStringSubmatch(clean); cm != nil {
			n = cm[1]
		}
		return fmt.Sprintf("ok: %s commits +%s-%s", n, stat[2], stat[3])
	}
	if pullCommitsRe.MatchString(clean) {
		return "ok: pulled"
	}
	return "ok"
}

var (
	commitLineRe        = regexp.MustCompile(`\[\S+\s+([0-9a-f]+)\]\s*(.*)`)
	commitNothingRe     = regexp.MustCompile(`(?i)nothing to commit`)
	branchCurrentLineRe = regexp.MustCompile(`^\*\s+(.+)`)
)

func compactCommit(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	if m := commitLineRe.FindStringSubmatch(clean); m != nil {
		if ultra {
			return "✓ " + shortHash7(m[1])
		}
		return "ok " + shortHash7(m[1]) + ": " + m[2]
	}
	if commitNothingRe.MatchString(clean) {
		if ultra {
			return "clean"
		}
		return "nothing to commit"
	}
	if ultra {
		return "✓"
	}
	return "ok"
}

func compactBranch(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	var lines []string
	for _, l := range strings.Split(clean, "\n") {
		if strings.TrimSpace(l) != "" {
			lines = append(lines, l)
		}
	}
	current := ""
	for _, line := range lines {
		if m := branchCurrentLineRe.FindStringSubmatch(line); m != nil {
			current = strings.TrimSpace(m[1])
		}
	}
	if ultra {
		return fmt.Sprintf("*%s (%db)", current, len(lines))
	}
	return fmt.Sprintf("current: %s | %d branches total", current, len(lines))
}
